app/src/main/java/de/danoeh/antennapod/activity/SingleFileAddActivity.java

```
@@ -0,0 +1,220 @@
+            return true;
+        }
+        return false;
+    }
+
+    private long getFileDuration(File f) {
+        MediaMetadataRetriever mediaMetadataRetriever = new MediaMetadataRetriever();
+        mediaMetadataRetriever.setDataSource(f.getAbsolutePath());
+        String durationStr = mediaMetadataRetriever.extractMetadata(MediaMetadataRetriever.METADATA_KEY_DURATION);
+        return Long.parseLong(durationStr);
+    }
+
+    /** Starts the import process. */
+    private void startImport(Uri uri) {
+        try {
+            String fileUrl = uri.toString();
+            Feed theDummyFeed = new Feed("someDummyUrl", null, "Dummy feed (local files)");
+
+            File f = new File(uri.getPath());
+
+            //create item
+            long globalId = 0;
+            FeedItem item = new FeedItem(globalId, "Dummy Feed: Item " + f.getName(), "item" + Long.toString(new Date().getTime()),
+                    fileUrl, new Date(), FeedItem.UNPLAYED, theDummyFeed);
+            item.setAutoDownload(false);
+
+            //add the media to the item
+            long duration = getFileDuration(f);
+            long size = f.length();
+            FeedMedia media = new FeedMedia(0, item, (int)duration, 0, size, "audio/mp3", f.getAbsolutePath(), f.getAbsolutePath(), true, null, 0, 0);
+            item.setMedia(media);
+
+            //add to the feed
+            List<FeedItem> items = new ArrayList<>();
+            items.add(item);
+            theDummyFeed.setItems(items);
+
+            //add or merge to the db
+            Feed[] feeds = DBTasks.updateFeed(this, theDummyFeed);
+
+        } catch (Exception e) {
+            Log.d(TAG, Log.getStackTraceString(e));
+            String message = getString(R.string.singlefile_import_error);
+            new MaterialDialog.Builder(this)
+                    .content(message + " " + e.getMessage())
+                    .positiveText(android.R.string.ok)
+                    .show();
```

I think it is better to throw the exception and display the dialog in the calling method. This allows to call `startImport` from a background service later (for automatic folder refresh).